File: server/common/lottery.py
import logging

from .utils import Bet, store_bets


class Lottery:
    # Parse the message and process the batch of bets.
    def handle_message(self, msg):
        fields = msg.split('\n')
        msg_type = fields[0]

        if msg_type != 'BATCH':
            logging.error(f'action: handle_message | result: fail | error: unknown message type {msg_type}')
            return 'ERR'

        agency = fields[1]
        count = int(fields[2])
        bets = []
        for i in range(count):
            parts = fields[3 + i].split(',')
            bets.append(Bet(agency, parts[0], parts[1], parts[2], parts[3], parts[4]))
        store_bets(bets)
        logging.info(f'action: apuesta_recibida | result: success | cantidad: {count}')
        return 'OK'

File: client/common/lottery.go
package common

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const MaxBatchBytes = 8 * 1024 // 8kiB

type Bet struct {
	Name      string
	Surname   string
	Document  string
	Birthdate string
	Number    string
}

// ReadBets reads all bets from a CSV file with format: name,surname,document,birthdate,number
func ReadBets(filePath string) ([]Bet, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open bets file: %w", err)
	}
	defer file.Close() // closes the file at the end of the function no matter the result.

	var bets []Bet
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		fields := strings.Split(scanner.Text(), ",")
		if len(fields) != 5 {
			return nil, fmt.Errorf("invalid bet line: %s", scanner.Text())
		}
		bets = append(bets, Bet{
			Name:      fields[0],
			Surname:   fields[1],
			Document:  fields[2],
			Birthdate: fields[3],
			Number:    fields[4],
		})
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("failed to read bets file: %w", err)
	}

	return bets, nil
}

// SerializeBatch builds the protocol message for a batch of bets.
// Format: BATCH\n{agency}\n{count}\n{name},{surname},{doc},{birth},{num}\n...
func SerializeBatch(agency string, bets []Bet) string {
	lines := []string{"BATCH", agency, fmt.Sprintf("%d", len(bets))}
	for _, b := range bets {
		lines = append(lines, fmt.Sprintf("%s,%s,%s,%s,%s", b.Name, b.Surname, b.Document, b.Birthdate, b.Number))
	}
	return strings.Join(lines, "\n")
}
